### PR TITLE
Wraedy/bigquery db connector

### PR DESCRIPTION
**Pull request by willyraedy**
Three main changes:
- Extends the BigQuery database connector to mimic more of the functionality of the Redshift connector.
- Extends the DB Connector class slightly.
- Moves BigQuery into the databases folder.

Main goal is to prepare the TMC network to migrate scripts from Redshift to BigQuery.

### CODE REVIEW

File: Dockerfile
Comment (willyraedy) [inline Dockerfile:1]: need to remove this before merge

File: .circleci/config.yml
Comment (willyraedy) [inline .circleci/config.yml:6]: I think a lot of these changes are because I merged "main" into my branch. Can we merge main into the "major-release" branch?

And apologies if that's not standard practice

File: Dockerfile
Comment (willyraedy) [inline Dockerfile:1]: Did the python support version get bumped to 3.8? If so, I'm confused why we didn't bump the image too

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [inline parsons/databases/bigquery/bigquery.py:1]: blarg. I messed up the file renaming the first time.

here's a commit where I tried to (unsuccessfully) restore the file diff if it's helpful: https://github.com/move-coop/parsons/commit/fc2c9fc915eae9e70595a67da4e4ba3369ef8c7c

File: parsons/databases/bigquery/__init__.py
Comment (willyraedy) [inline parsons/databases/bigquery/__init__.py:1]: Went back and forth on which organizational convention to follow. Decided to move to the databases folder but no strong opinion here if ppl want to leave in the google folder

File: parsons/google/google_cloud_storage.py
Comment (willyraedy) [inline parsons/google/google_cloud_storage.py:151]: TODO: cut this

File: parsons/databases/postgres/postgres.py
Comment (willyraedy) [inline parsons/databases/postgres/postgres.py:83]: woops

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [inline parsons/databases/bigquery/bigquery.py:177]: Is this connection even worth having on the BigQuery connector because of the state issues?

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [inline parsons/databases/bigquery/bigquery.py:275]: Kept this param to match with Redshift and then throw with a descriptive error. Could just omit the param altogether

**Comment by willyraedy** [comment]
@shaunagm @Jason94 I'm gonna add some unit tests before I mark this ready to merge but if y'all want to give it a once over should be ready for that

**Comment by sharinetmc** [comment]
@shaunagm Hiya! Just wondering if you had the time to take a look at this!

File: .circleci/config.yml
Comment (Jason94) [reply .circleci/config.yml:6]: Yeah, it's not super well defined at the moment. But it should be safe for anyone to merge main into major-release whenever they need to.

Comment (willyraedy) [reply .circleci/config.yml:6]: Sure

Comment (willyraedy) [reply .circleci/config.yml]: Ok should be g2g

File: parsons/databases/bigquery/__init__.py
Comment (Jason94) [reply parsons/databases/bigquery/__init__.py:1]: I prefer this way. I think it makes more sense to be in databases.

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:1084]: What is the point of doing the pickling 100,000 at a time if we're just going to load them all back into memory at the end anyway?

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:327]: This type annotation should be `Optional[str]`

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:328]: This type annotation should be `Optional[list]`, or better yet `Optional[List[dict]]`

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:319]: Feel free to not do this (don't want to scope creep your PR too much) but it would be better to use an enum for this instead of a string.

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:423]: Maybe these could be debug log level?

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:468]: Same comments about file type

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:806]: Could we link to the BQ copy method instead?

File: parsons/databases/bigquery/bigquery.py
Comment (Jason94) [inline parsons/databases/bigquery/bigquery.py:1111]: I don't know if this is out of date or what, but it looks like it's a simple command now: https://cloud.google.com/bigquery/docs/reference/standard-sql/dml-syntax#truncate_table_statement

**Review by Jason94 — CHANGES_REQUESTED** [review]
I left comments about the log level in a few places, but didn't touch it every place it could've been changed.

In general I think Parsons over-uses the info log level. I think that the info log level should be used if it's information that I might care about if I need to look back on what was happening in my script. For example, if I have a script that is uploading some new contacts, I probably want to have a log of the event ID's that were generated. But I probably don't care that it was compressing the table data as part of the GCS upload, or whatnot.

Something like that could be justified at the debug log level, maybe. The reason being that if something fails then it might be helpful to change the log level to DEBUG to see under the hood of what's happening. But even then, I'm not sure if giving a "Yep, this function is running." style of log message (that gives no information) is helpful.

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [reply parsons/databases/bigquery/bigquery.py:1084]: This was inherited from the previous version of BigQuery. Honestly didn't interrogate it much for improvements and I'm inclined to leave it for now

**Comment by willyraedy** [comment]
Appreciate it Jason! I'm a little sick so @IanRFerguson is gonna take a run at making those changes. They mostly seem pretty minor so I suspect he'll be able get through those today or tomorrow

File: parsons/databases/bigquery/bigquery.py
Comment (IanRFerguson) [reply parsons/databases/bigquery/bigquery.py:1111]: @willyraedy didn't address this one in my review, let's connect on it

**Comment by IanRFerguson** [comment]
@Jason94 I used some discretion to move many of the info-level logs to debug ... there's one outstanding comment that I'm talking through with Willy now but then we should be all set for another review

**Review by shaunagm — CHANGES_REQUESTED** [review]
Overall this looks good to me - some minor nitpicks and flags. The main question I have is around the deprecation of the old connector. 

Also note that I have no personal experience with the BigQuery connector so my review is largely for style and readability, not for substance.

File: parsons/__init__.py
Comment (willyraedy) [reply parsons/__init__.py:56]: I guess I was thinking it would be part of a major release so breaking changes would be expected. But now that I think about that, I don't think there's any reason we can't import it from both places.

Comment (willyraedy) [reply parsons/__init__.py:56]: Moved the file back to the original location and modified the database init file to also import it from there. So both should work

Comment (shaunagm) [reply parsons/__init__.py:56]: Thanks for making this adjustment. Where possible, we want to give folks time to make the change, so in 3-6 we can circle back and remove this entirely.

The docs should signal this deprecation. I missed that this lacks a docs update in my first review, but this should update the docs so that the BigQuery docs are in the Databases folder instead of in the Google folder, both in the git repository and in the docs lefthand sidebar. It probably makes sense to leave a stub in the Google folder, though, that says something like "GoogleBigQuery has been moved to the Databases folder and is available under BigQuery, please change this in your code if you don't want it to break later".

Comment (willyraedy) [reply parsons/__init__.py:56]: I think the big query docs can still live in the same place since I moved the file back to its original place. Unless I'm missing something.

All we've basically done is take the original class and export it in the database package as well

Comment (shaunagm) [reply parsons/__init__.py:56]: I double-checked the docs and I was misremembering how they're structured. There's actually no subfolders at all, including no databases folder. Whoops. We can keep the databases .rst file as is - like how Redshift points to the entry under AWS, BigQuery's spot in the databases listing can keep pointing to Google.

That said, we do still want a note about the rename and the planned deprecation in the docs, so that anyone who checks the docs over the next 3-6 months will be able to proactively rename, and then when we finally do make the breaking change of removing GoogleBigQuery as an option, fewer people will be affected.

Comment (willyraedy) [reply parsons/__init__.py]: The more I think about it, I don't think renaming it from GoogleBigQuery to BigQuery which is functionally what it would be to users would be worth a deprecation process. So I just removed the second import path and will change our code.

Hoping to get this merged this week so we can onboard a contractor on Monday!

**Review by Jason94 — APPROVED** [review]

File: parsons/databases/bigquery/bigquery.py
Comment (coastlines) [reply parsons/databases/bigquery/bigquery.py]: Bumping this because I'm not sure if it was ever resolved.

File: parsons/databases/bigquery/bigquery.py
Comment (coastlines) [reply parsons/databases/bigquery/bigquery.py]: @Jason94 can you say more on why this would be beneficial? Purely out of curiosity, no rush.

File: parsons/databases/bigquery/bigquery.py
Comment (coastlines) [inline parsons/databases/bigquery/bigquery.py]: Update to`bigquery.connection()`

File: parsons/databases/bigquery/bigquery.py
Comment (coastlines) [inline parsons/databases/bigquery/bigquery.py]: Should this be referencing the BigQuery method?

Comment (coastlines) [reply parsons/databases/bigquery/bigquery.py]: Sorry, just saw Jason's comment. It seems this has been resolved.

File: parsons/etl/tofrom.py
Comment (coastlines) [reply parsons/etl/tofrom.py]: This confuses me too so I'm unresolving this conversation. It would be super helpful to have the documentation for this updated before merging.

File: parsons/etl/tofrom.py
Comment (coastlines) [inline parsons/etl/tofrom.py:949]: Yes but not urgent. Passing in kwargs is often very helpful for strategists.

**Review by coastlines — APPROVED** [review]
Impressive work y'all!

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [reply parsons/databases/bigquery/bigquery.py]: good catch

File: parsons/databases/bigquery/bigquery.py
Comment (willyraedy) [reply parsons/databases/bigquery/bigquery.py]: I think given the error checks we've put in, it's probably not worth deprecating since this was already on the class before

File: parsons/etl/tofrom.py
Comment (willyraedy) [reply parsons/etl/tofrom.py]: this was updated! looks like the Github diff is cached somehow

**Pull request merged**
